pkg/logging/logging.go

```diff
@@ -0,0 +1,163 @@
+		filterLevel: defaultLogLevel,
+		currentLogLevel: defaultLogLevel,
+	}
+}
+
+type NullLogger struct{}
+func (n NullLogger) Log(params ...interface{}) error { return nil }
+
+var loggers =  make(map[string]*FilteredLogger)
+
+func Logger(component string) *FilteredLogger {
+	if _, ok := loggers[component]; !ok {
+		logger := log.NewLogfmtLogger(os.Stderr)
+		log := MakeLogger(logger)
+		log.component = component
+		loggers[component] = log
+	}
+	return loggers[component]
+}
+
+func (l *FilteredLogger) SetLogger(logger log.Logger) *FilteredLogger {
+	l.logger = logger
+	return l
+}
+
+var logFilterLevels = map[string]int {
+	"debug":    0,
+	"info":     10,
+	"warning":  20,
+	"error":    30,
+	"critical": 40}
+
+type LogError struct {
+	message string
+}
+
+func (e LogError) Error() string {
+	return e.message
+}
+
+func (l FilteredLogger) Log(params ...interface{}) error {
+	if (l.currentLogLevel >= l.filterLevel) && (l.currentVerbosityLevel <= l.verbosityLevel) {
+		logParams := make([]interface{}, 0)
+		logParams = append(logParams, "component", l.component)
+
+		if l.currentVerbosityLevel > 0 {
+			_, fileName, lineNumber, _ := runtime.Caller(0)
+			logParams = append(logParams, "filename", fileName)
+			logParams = append(logParams, "line", lineNumber)
+		}
+		logParams = append(logParams, params...)
+		return l.logger.Log(logParams...)
+
+	}
+	return nil
+}
+
+func (l FilteredLogger) Object(obj ...interface{}) error {
+	// name, ID, type
+	t := reflect.TypeOf(obj).Elem()
+	logParams := make([]interface{}, 0)
+	logParams = append(logParams, "component", l.component)
+
+	if l.currentVerbosityLevel > 0 {
+		_, fileName, lineNumber, _ := runtime.Caller(1)
+		logParams = append(logParams, "filename", fileName)
+		logParams = append(logParams, "line", lineNumber)
+	}
+
+	logParams = append(logParams, "type", t.String())
+	logParams = append(logParams, "name", t.Name())
+
+	if (t.Kind() == reflect.Array) || (t.Kind() == reflect.Slice) {
+		logParams = append(logParams, "length", t.Len())
+	}
+	if t.Kind() == reflect.Struct {
+		logParams = append(logParams, "fields", t.NumField())
+	}
+
+	return l.logger.Log(logParams)
+}
+
+
+
+func (l *FilteredLogger) SetLogLevel(filter string) error {
+	for filterName, filterLevel := range logFilterLevels {
+		if filter == filterName {
+			l.filterLevel = filterLevel
+			return nil
+		}
+	}
+	return errors.New("Log level " + filter + " does not exist")
+}
+
+func (l *FilteredLogger) SetVerbosityLevel(level int) error {
+	if level >= 0 {
+		l.verbosityLevel = level
+	} else {
+		return errors.New("Verbosity setting must not be negative")
+	}
+	return nil
+}
+
+// It would be consistent to return an error from this function, but
+// a multi-value function would break the primary use case: log.V(2).Info()....
+func (l FilteredLogger) V(level int) *FilteredLogger {
+	if level >= 0 {
+		l.currentVerbosityLevel = level
+	}
+	return &l
+}
+
+func (l FilteredLogger) Debug() *FilteredLogger {
```

Comment:
With V() for info level, we don't need the Debug level anymore:

I would think about the following usage:

```golang
func test() {
    log.Error().Log()
    log.V(2).Info() // When we do V(), only Info() is left
    log.Info() // Could have V(1) implicitly
   log.Warning() // No way to set V(1) before or after Warning
} 
```

Since we then have no Debug level anymore, we don't need any filter logic by log message type. There is never any reason not to show Critical, Error, Warning and V(1).Info. With additional -vvv we would also log V(2).Info, V(3).Info.